packages/ove-service-layout/src/validator/extensions.js

```diff
@@ -0,0 +1,71 @@
+const validate = require("validate.js");
+
+validate.validators.isIn = (value, options) => options.includes(value) ? null : `should be one of [${options}]`;
+validate.validators.isEqual = (value, options) => options === value ? null : `shouldn't be equal to ${options}`;
+validate.validators.isNotEqual = (value, options) => options !== value ? null : `shouldn't be equal to ${options}`;
+validate.validators.isNotEmpty = (value) => validate.isEmpty(value) ? `can't be empty` : null;
+validate.validators.isNumber = (value) => validate.isNumber(value) ? null : `should be a number`;
+validate.validators.isPercent = (value) => validate.isNumber(value) && value >= 0 && value <= 1 ? null : `should be between [0, 1]`;
+validate.validators.isString = (value) => validate.isString(value) ? null : `should be a valid string`;
+
+const translateValidator = (options, containers) => {
+    let result = {};
+    for (let k of Object.keys(options)) {
+        for (let container of containers) {
+            result[k.replace("#", container)] = options[k];
+        }
+    }
+    return result;
+};
+
+const sectionValidator = (sections, options) => {
+    let errors = [];
+    if (sections) {
+        for (let section of sections) {
+            if (!validate.isEmpty(section.name) && validate.isString(section.name)) {
+                let vErrors = validate({[section.name]: section}, translateValidator(options, [section.name]));
+                if (vErrors) {
+                    errors = [...errors, vErrors];
+                }
+
+                if (section.type === "container" && !validate.isEmpty(section.sections)) {
+                    vErrors = sectionValidator(section.sections, options);
+                    if (vErrors) {
+                        errors = [...errors, ...vErrors];
+                    }
+                }
+            } else {
+                errors.push("One of the sections has an empty or non-string name");
+            }
+        }
+    }
+    return errors.length > 0 ? errors : null;
+};
+
+validate.validators.sectionValidator = sectionValidator;
+
+validate.validators.containerValidator = (container) => {
+    let errors = [];
+    if (container.type === "container") {
+        if (validate.isEmpty(container.sections)) {
+            errors.push("sections are mandatory for containers");
+        }
+
+        if (validate.isEmpty(container.layout)) {
+            errors.push("layout property is mandatory for containers");
+        } else if (validate.isEmpty(container.layout.type) || !validate.isString(container.layout.type)) {
+            errors.push("layout type needs to be a valid non-empty string for containers");
+        }
+    } else if (container.type === "section") {
+        if (container.sections) {
+            errors.push("sections can't have subsections");
+        }
+
+        if (container.layout) {
+            errors.push("sections can't have layouts, please upgrade to container class");
```

Comment:
```suggestion
            errors.push("sections can't have layouts, please use a container class");
```

Reply:
I actually like the original message ...

Reply:
Depends on how the UI works. If the UI lets you upgrade a section into a container, then this is fine. If it requires you to create a new container and then place the section in it, then this does not make sense. The suggestion is more neutral, but I don't mind either.

Reply:
I don't think a container really is an "upgrade" of a section: it's a different `thing.` I suggest either:

```suggestion
            errors.push("sections can't have layouts: please use a container class instead");
```

or

```suggestion
            errors.push("sections can't have layouts: please replace with a container class");
```

Reply:
The original message was meant as a joke, but I changed it to James' suggestion